clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -119,6 +123,11 @@ public Fetcher(ConsumerNetworkClient client,
      * @param cluster The current cluster metadata
      */
     public void initFetches(Cluster cluster) {
+        // Don't initiate fetch if we already have records available from previous fetchRequest
+        // This may happen when there is OffsetOutOfRange exception from previous fetchRequest
+        if (!records.isEmpty())
```

Comment:
For 1), consider this sequence:

consumer.poll(100); // fetches sent here
consumer.commit(SYNC); // fetches returned here
consumer.poll(100); // should return immediately, but will block on the next round of fetches because we don't check for available records before polling

For 2), I'm a little unclear why you seem reluctant to change pollOnce. I'm pretty sure that the only place it's used is in KafkaConsumer.poll().